Fix ReactiveDependencies rule

jamiebuilds opened the pull request:
I'm interested in how parts of this might have been working in the past because it seemed to rely on things that ESLint does not do.

In particular, ESLint doesn't have a NodePath type like Babel so they assign `node.parent` as the parent node. But for some reason they don't do this everywhere. So you end up in scenarios like we had here where you don't have access to the parent node.

I was not able to find a way to get the parent node from ESLint's `context.getScope().references`. So I was forced to do the worst possible thing and search through the entire AST for the node we're looking for.

I've optimized this search as much as possible. It's similar to a binary search except we're working with a generic tree. The search doesn't know anything about the AST though, it just looks for anything with a `{ type: string, ... }` and assumes that has the same basic interface as other nodes do in ESLint.

calebmer committed:
Add ESLint rule for useEffect/useCallback/useMemo Hook dependencies

ljharb committed:
[react-is] add back proper AsyncMode symbol, for back compat (#13959)

- Partial revert of #13732
 - Fixes #13958.

jamiebuilds committed:
Fix ReactiveDependencies rule

jamiebuilds commented:
@gaearon I'm assuming this PR isn't going to run CI since it's in a different repo... What do you want me to do with it?

gaearon commented:
Want to just submit both original commit and further work on top as your own PR? I’ll close mine. Let’s just keep first commit credited to Caleb.

jamiebuilds committed:
fix lint errors

mmarkelov committed:
Simplify imports in react reconciler (#13718)

* Simplify imports in ReactChildFiber
* Import type first in ReactCurrentFiber
* Simplify imports in ReactFiberBeginWork
* Simplify imports in ReactFiberScheduler
* Simplify import in ReactFiberTreeReflection
* Simplify import in ReactFiberUnwindWork
* Remove repeated import
* Fix imports from ReactFiberExpirationTime
* Master imports in ReactFiberBeginWork

chrisbianca committed:
Remove duplicate import declarations (#14062)

sophiebits committed:
Tweak error message for missing fallback (#14068)

sophiebits committed:
Warn for forwardRef(memo(...)) (#14070)

People are probably gonna do this all the time.

Andarist committed:
Rename inputsAreEqual to areHookInputsEqual & move it to shared (#14036)

sophiebits committed:
Warn for bad useEffect return value (#14069)

Mostly to catch this:

```js
useEffect(async () => {
  // ...
  return cleanup;
});
```

Is this too restrictive? Not sure if you would want to do like

```js
useEffect(() => ref.current.style.color = 'red');
```

which would give a false positive here. We can always relax it to only warn on Promises if people complain.

SimenB committed:
chore(tests): don't rely on jest fake timers scheduling real timers (#14003)

* chore: don't rely on jest fake timers scheduling real timers

* re-add one part not working with Jest 23

einarq committed:
Bugfix: Add back early return in setOffsets which was removed between 16.4.2 and 16.5.0. Fails in Edge in some scenarios. (#14095)

marcelometal committed:
react-reconciler: Update README.md (#13953)

Replaced 'var' with 'const'

keyz committed:
update the benchmark script (#13994)

tnunes committed:
fix(react-dom): Fix crash during server render (#14103)

Check for existence of `setTimeout` and `clearTimeout` in the runtime
before using them, to ensure runtimes without them (like .NET ClearScript)
do not crash just by importing `react-dom`.

sebmarkbage committed:
Add Debug Tools Package for Introspection of Hooks (#14085)

* Add debug tools package

* Add basic implementation

* Implement inspection of the current state of hooks using the fiber tree

* Support useContext hooks inspection by backtracking from the Fiber

I'm not sure this is safe because the return fibers may not be current
but close enough and it's fast.

We use this to set up the current values of the providers.

* rm copypasta

* Use lastIndexOf

Just in case. I don't know of any scenario where this can happen.

* Support ForwardRef

* Add test for memo and custom hooks

* Support defaultProps resolution

sophiebits committed:
Use Function.prototype.apply in warningWithoutStack (#14107)

console.error.apply() fails in IE9, but I verified this works (and it works everywhere else too). :)

sophiebits committed:
Rename .internal tests that aren't using internals (#14109)

sophiebits committed:
DCE hooks code when flag is off (#14111)

sophiebits committed:
Implement {,un}hideInstance on RN renderer (#14115)

This is required to use lazy.

Test Plan:
* Verified lazy works on a real world use case (shows spinner, shows real content).
* Verified that if I change the primary content's styles to have `display: 'none'` then it never appears (i.e., the code in `unhide` reads the styles successfully)

acdlite committed:
[suspense] Avoid double commit by re-rendering immediately and reusing primary children (#14083)

* Avoid double commit by re-rendering immediately and reusing children

To support Suspense outside of concurrent mode, any component that
starts rendering must commit synchronously without being interrupted.
This means normal path, where we unwind the stack and try again from the
nearest Suspense boundary, won't work.

We used to have a special case where we commit the suspended tree in an
incomplete state. Then, in a subsequent commit, we re-render using the
fallback.

The first part — committing an incomplete tree — hasn't changed with
this PR. But I've changed the second part — now we render the fallback
children immediately, within the same commit.

* Add a failing test for remounting fallback in sync mode

* Add failing test for stuck Suspense fallback

* Toggle visibility of Suspense children in mutation phase, not layout

If parent reads visibility of children in a lifecycle, they should have
already updated.

gaearon committed:
Restore the Hooks dispatcher after using SSR (#14105)

acdlite committed:
Enable hooks in www build (#14116)

The `enableHooks` feature flag used to only control whether the API
was exposed on the React package. But now it also determines if the
dispatcher and implementation are included in the bundle.

We're using hooks in www, so I've switched the feature flag to `true`
in the www build.

(Alternatively, we could have two feature flags: one for the
implementation and dispatcher, and one for exposing the API on the
React package.)

acdlite committed:
Enable hooks in www test renderer, too

sophiebits committed:
Fix unhiding in IE11 (#14126)

Setting to null isn't correct; setting to '' is. I opted to use dangerousStyleValue for consistency with the main path that we set things.

Fixes #14114.

Test Plan:
Verified setting to '' works in Chrome and IE11. (Setting to null works in Chrome but not in IE11.)

gaearon committed:
Fix lazy() with defaultProps (#14112)

* Resolve defaultProps for Lazy components

* Make test fail again

* Undo the partial fix

* Make test output more compact

* Add a separate failing test for sync mode

* Clean up tests

* Add another update to both tests

* Resolve props for commit phase lifecycles

* Resolve prevProps for begin phase lifecycles

* Resolve prevProps for pre-commit lifecycles

* Only resolve props if element type differs

* Fix Flow

* Don't set instance.props/state during commit phase

This is an optimization. I'm not sure it's entirely safe. It's probably worth running internal tests and see if we can ever trigger a case where they're different.

This can mess with resuming.

* Keep setting instance.props/state before unmounting

This reverts part of the previous commit. It broke a test that verifies we use current props in componentWillUnmount if the fiber unmounts due to an error.

gaearon committed:
Add DEV-only checks for assumption about instance properties (#14128)

acdlite committed:
Check correct commit phase props in fuzz tester (#14129)

Adds a check to the existing fuzz tester to confirm that the props are
set to the latest values in the commit phase. Only checks
componentDidUpdate; we already have unit tests for the other lifecycles,
so I think this is good enough. This is only a redundancy.

gaearon committed:
Recover from errors with a boundary in completion phase (#14104)

* Recover from errors with a boundary in completion phase

* Use a separate field for completing unit of work

* Use a simpler fix with one boolean

* Reoder conditions

* Clarify which paths are DEV-only

* Move duplicated line out

* Make it clearer this code is DEV-only

nadavkaner committed:
Remove unused simulated flag parameter (#14127)

locknono committed:
"functional component" -> "function component" (#14123)

acdlite committed:
Make react-debug-tools a private package

acdlite committed:
Update changelog with unreleased features

acdlite committed:
Updating yarn.lock file for 16.6.1 release

acdlite committed:
Updating package versions for release 16.6.1

acdlite committed:
Update error codes for 16.6.1 release

acdlite committed:
Update bundle sizes for 16.6.1 release

acdlite committed:
Updating CHANGELOG.md for 16.6.1 release

acdlite committed:
Make react-debug-tools a private package

trueadm committed:
Show deprecated context object warnings usage in ReactDOM server (#14033)

* Applies context object warnings to ReactDOM server

acdlite committed:
[Synchronous Suspense] Reuse deletions from primary tree (#14133)

Fixes a bug where deletion effects in the primary tree were dropped
before entering the second render pass.

Because we no longer reset the effect list after the first render pass,
I've also moved the deletion of the fallback children to the complete
phase, after the tree successfully renders without suspending.

Will need to revisit this heuristic when we implement resuming.

bvaughn committed:
Release script supports interleaved stable and alpha releases (#14138)

bvaughn committed:
Tag MemoComponent with PerformedWork effectTag for DevTools Profiler (#14141)

bvaughn committed:
Invalid actualDuration+treeBaseDuration for hidden+suspended trees (#14065)

* Fixed `treeBaseDuration` by propagating its value from the suspended tree to the Fragment React temporarily wraps around it when showing the fallback UI.
* Fixed `actualDuration` by recording elapsed profiler time in the event of an error.
* Fixed `actualDuration` in concurrent mode by propagating the time spent rendering the suspending component to its parent.

Also updated ReactSuspensePlaceholder-test.internal to cover these new cases.

sebmarkbage committed:
Use Entry in `yarn build ...` Instead of Label (#14148)

* Parse build script type and package names

This ensures that `yarn build core dom` includes DOM.

It also ensures that spaces like `yarn build "core, dom"` doesn't build EVERYTHING.

* Get rid of label in bundles config

Instead we just use the name from entry using fuzzy search.

There is one special case. If you put in `/index` or `/index.js`.

That allows to build things like `react/index` to only build isomorphic
where as `react` would build everything. Or `react-dom/index` to exclude
the server renderers.

* Instead of matching `/index.js` just append it to the search string

That way things like `yarn build react/` works too.

NMinhNguyen committed:
Refactor ESLint configuration to enable better IDE integration (#13914)

* Refactor ESLint configuration to enable better IDE integration

* Minor tweaks

n8schloss committed:
Adding logger pri (#14155)

acdlite committed:
[Synchronous Suspense] Don't delete children of suspended component (#14157)

Vestigial behavior that should have been removed in #13823.

Found using the Suspense fuzz tester in #14147.

acdlite committed:
Don't warn if an unmounted component is pinged (#14158)

* Add failing test for ping on unmounted component

We had a test for this, but not outside of concurrent mode :)

* Don't warn if an unmounted component is pinged

acdlite committed:
Suspense fuzz tester (#14147)

* Don't warn if an unmounted component is pinged

* Suspense fuzz tester

The fuzzer works by generating a random tree of React elements. The tree
two types of custom components:

- A Text component suspends rendering on initial mount for a fuzzy
  duration of time. It may update a fuzzy number of times; each update
  supsends for a fuzzy duration of time.
- A Container component wraps some children. It may remount its children
  a fuzzy number of times, by updating its key.

The tree may also include nested Suspense components.

After this tree is generated, the tester sets a flag to temporarily
disable Text components from suspending. The tree is rendered
synchronously. The output of this render is the expected output.

Then the tester flips the flag back to enable suspending. It renders the
tree again. This time the Text components will suspend for the amount of
time configured by the props. The tester waits until everything has
resolved. The resolved output is then compared to the expected output
generated in the previous step.

Finally, we render once more, but this time in concurrent mode. Once
again, the resolved output is compared to the expected output.

I tested by commenting out various parts of the Suspense implementation
to see if broke in the expected way. I also confirmed that it would have
caught #14133, a recent bug related to deletions.

* When a generated test case fails, log its input

* Moar fuzziness

Adds more fuzziness to the generated tests. Specifcally, introduces
nested Suspense cases, where the fallback of a Suspense component
also suspends.

This flushed out a bug (yay!) whose test case I've hard coded.

* Use seeded random number generator

So if there's a failure, we can bisect.

acdlite committed:
Remove errant return assignment (#14164)

Oopsie!

This could have been avoided if our types were modeled correctly with
Flow (using a disjoint union).

Fuzz tester didn't catch it because it does not generate cases where
a Suspense component mounts with no children. I'll update it.

alexmckenley committed:
Improved suspense support in ReactDOMServer (#14161)

acdlite committed:
Add fb build of ESLint plugin (#14165)

acdlite committed:
Add global to ESLint plugin bundle config

bgord committed:
Fix comment typo (#14156)

NE-SmallTown committed:
Fix typos (#14124)

sophiebits committed:
Don't run danger on bad build (#14143)

sizebot comments can be confusing when not based on reality.

If results.json doesn't exist, danger will fail. This is what we want.

acdlite committed:
ESlint -> ESLint

bvaughn committed:
Renamed outdated schedule/tracing referecnes (#14177)

sophiebits committed:
Warn about conflicting style values during updates (#14181)

This is one of the most insidious quirks of React DOM that people run into. Now we warn when we think an update is dangerous.

We still allow rendering `{background, backgroundSize}` with unchanging values, for example. But once you remove either one or change `background` (without changing `backgroundSize` at the same time), that's bad news. So we warn.

Fixes #6348.

sophiebits committed:
SimpleMemoComponent should warn if a ref is given (#14178)

Fixes #13964.

sebmarkbage committed:
Use unique thread ID for each partial render to access Context (#14182)

* BUG: ReactPartialRenderer / New Context polutes mutable global state

The new context API stores the provided values on the shared context instance. When used in a synchronous context, this is not an issue. However when used in an concurrent context this can cause a "push provider" from one react render to have an effect on an unrelated concurrent react render.

I've encountered this bug in production when using renderToNodeStream, which asks ReactPartialRenderer for bytes up to a high water mark before yielding. If two Node Streams are created and read from in parallel, the state of one can polute the other.

I wrote a failing test to illustrate the conditions under which this happens.

I'm also concerned that the experimental concurrent/async React rendering on the client could suffer from the same issue.

* Use unique thread ID for each partial render to access Context

This first adds an allocator that keeps track of a unique ThreadID index
for each currently executing partial renderer. IDs are not just growing
but are reused as streams are destroyed.

This ensures that IDs are kept nice and compact.

This lets us use an "array" for each Context object to store the current
values. The look up for these are fast because they're just looking up
an offset in a tightly packed "array".

I don't use an actual Array object to store the values. Instead, I rely
on that VMs (notably V8) treat storage of numeric index property access
as a separate "elements" allocation.

This lets us avoid an extra indirection.

However, we must ensure that these arrays are not holey to preserve this
feature.

To do that I store the _threadCount on each context (effectively it takes
the place of the .length property on an array).

This lets us first validate that the context has enough slots before we
access the slot. If not, we fill in the slots with the default value.

sebmarkbage committed:
fix typo

sophiebits committed:
Simplify CSS shorthand property warning (#14183)

I figured out a simpler way to do #14181. It does allocate some but I think that's OK. Time complexity might even be better since we avoid the nested loops the old one had.

mmarkelov committed:
Update Readme (#14176)

Replace core with react, dom-client with react-dom

acdlite committed:
Update error codes

gaearon committed:
Add regression test for #14188 (#14197)

bvaughn committed:
Save CI-built node_modules as build artifacts (#14205)

* Store node_modules generated by CI script as an artifact
* NPM pack artifacts before archiving

gaearon committed:
Add 16.6.3 Changelog (#14223)

* Add 16.6.3 Changelog

* Remove unreleased fix

* Drop another unreleased fix

acdlite committed:
[scheduler] Post to MessageChannel instead of window (#14234)

Scheduler needs to schedule a task that fires after paint. To do this,
it currently posts a message event to `window`. This happens on every
frame until the queue is empty. An unfortunate consequence is that every
other message event handler also gets called on every frame; even if
they exit immediately, this adds up to significant per-frame overhead.

Instead, we'll create a MessageChannel and post to that, with a
fallback to the old behavior if MessageChannel does not exist.

acdlite committed:
[scheduler] Remove window.postMessage fallback

Every browser we can about supports MessageChannel. The ones we don't
care about will fallback to the setTimeout implementation.

poeschko committed:
Add a checkbox to fixtures UI to choose React production build (#13786)

* Add a checkbox to fixtures UI to choose React production build

* Assign header__label class name to label directly, instead of using a separate span

* center the production checkbox vertically

acdlite committed:
Wrap shorthand CSS property collision warning in feature flag (#14245)

Disables the recently introduced (#14181) warning for shorthand
CSS property collisions by wrapping in a feature flag. Let's hold off
shipping this until at least the next minor.

developit committed:
Reduce scheduler serialization overhead (#14249)

In the process of switching to MessageChannel, it seems the postMessage call was modified to pass `"*"` (originally the target origin value from `window.postMessage`). This actually ends up triggering serialization, whereas passing `undefined` bypasses.

To save some investigation, passing a Number like `0` still incurs serialization overhead - `undefined` has special behavior.

trueadm committed:
Clear memoizedState on unmount of fiber to avoid memory leak (#14218)

* Clear properties on unmount of fiber to ensure objects are not retained

trueadm committed:
Revert "Clear memoizedState on unmount of fiber to avoid memory leak (#14218)" (#14275)

This reverts commit 9b2fb24f99ebed9af2c092de7a5f31c38f8af97a.

trueadm committed:
Clear fields on unmount of fiber to avoid memory leak (#14276)

* Clear fields on unmount of fiber to avoid memory leak

gaearon committed:
Fix SSR useCallback in render phase (#14279)

sebmarkbage committed:
Remove `import * as` pattern from the codebase (#14282)

Whenever we do this, Rollup needs to materialize this as an object.
This causes it to also add the Babel compatibility property which is
unnecessary bloat. However, since when we use these, we leak the object
this often also deopts any compiler optimizations.

If we really need an object we should export default an object.

Currently there is an exception for DOMTopLevelEventTypes since
listing out the imports is a PITA and it doesn't escape so it should
get properly inlined. We should probably move to a different pattern
to avoid this for consistency though.

gaearon committed:
Fix support for mixing react-dom/server@16.6 and react@<16.6 (#14291)

gaearon committed:
Warn about reassigning this.props (#14277)

* Warn about reassigning this.props

* Improve the warning

* Don't show the spammy bug warning if we suspect it's a component bug

Commit:
fix spelling error: Here's -> Here (#14307)

"Here's" should be changed to "Here" in the given sentence.

gaearon committed:
List ignored types instead of included types in the stack (#14308)

gaearon committed:
Use |0 to coerce to number (#14297)

gaearon committed:
Add more test coverage for nested memo() (#14311)

gaearon committed:
Fix resolution of outer props with React.memo() (#14312)

* Add failing test for defaultProps between lazy() and memo()

* Add another regression test for defaultProps resolution order

* Resolve outer props for MemoComponent

gaearon committed:
Prevent deopts from modifying exports object in stable builds (#14309)

bvaughn committed:
Publish a local release (canary or stable) to NPM (#14260)

New release scripts.

Learn more at https://github.com/facebook/react/blob/master/scripts/release/README.md

bvaughn committed:
Add basic release script snapshot test (#14280)

Added regression test for release scripts

Commit:
update fixtures/packaging/README.md (#14320)

I had some confusion yesterday with this, figured this reads a little better.  Ran these steps locally to verify they work.

bvaughn committed:
Add progress bars to longer running async release tasks (#14322)

* Add progress bars to longer running async release tasks
* Updated to 0.2 progress estimator version

cpojer committed:
Remove usage of `fbjs/lib/invariant` in ReactNativeViewConfigRegistry. (#14330)

voxpelli committed:
Fix regression: Errors not emitted in streams (#14314)

Regression introduced in #14182 resulted in errors no longer being emitted on streams, breaking many consumers.

Co-authored-by: Elliot Jalgard <elliot.j@live.se>

gaearon committed:
Make useEffect(async) warning more verbose (#14327)

* Make useEffect(async) warning more verbose

* Nit

springuper committed:
Add a null type test for memo (#14325)

sophiebits committed:
Remove useMutationEffect (#14336)

useMutationEffect has problems (namely, refs aren't attached at the time that it runs) and we're not positive it's necessary. useLayoutEffect runs at the same time as componentDidMount/Update so it's sufficient for all existing use cases; it can be used in any case that useEffect happens too late. Until we figure out what we want to do, let's delete it.

jintoppy committed:
Adding isMemo check to react-is package (#14313)

bvaughn committed:
Release script auto-determine the latest Canary build if none specified (#14339)

* Release script auto-determine the latest Canary build if none specified

gaearon committed:
Validate propTypes for lazy() and memo() and warn about invalid patterns (#14298)

* Add a test for current defaultProps behavior in lazy

* Add a warning against definining defaultProps on the outer wrapper

* Warn about setting propTypes too

* Remove redundant async

* Validate propTypes for resolved lazy types

Note this only works for elements created after resolving. So it's not ideal. But it provides the best stack trace for those cases.

* Add a test for lazy(forwardRef()) propTypes check

* Validate memo() inner propTypes and warn about shadowing

* Add test verifying nested lazy is unsupported

* Change error wording to remove "Promise elements"

* Improve error message for nested lazy() and add tests

* Validate propTypes for memo in the reconciler when necessary

* Add comments for why we're calling checkPropTypes

* Fix Flow and lint

* Undo unintentional formatting changes

* Remove unnecessary case (it is handled by function code path)

* Add test coverage for memo(fn.defaultProps).propTypes

* Test should be agnostic of where resolving happens

That's an implementation detail and we might want to change it later. Let's keep it easy by making tests just check that validation happened, not at which stage.

* Unify traversal logic in createElement

This moves all type traversal into createElement. When lazy resolves, we call createElement once to re-check.

* Match prod behavior for propTypes/defaultProps shims closer

* Revert "Unify traversal logic in createElement"

This reverts commit 2e77ca47fe80ebe6595333542a8c5c138c68643f.

See https://github.com/facebook/react/pull/14298#issuecomment-442687775

* Undo unnecessary change to getComponentName

gaearon committed:
[Fire] Add initial build infrastructure (#14359)

ioss committed:
Fix bug in cloneHook (#14364)

* Fixes #14360 and adds a test for mixed priority dispatches.

It was broken because `cloneHook` assigned `memoizedState` instead of
`baseState` from the original hook to `baseState` of the clone.

* tweak comments

sebmarkbage committed:
[Fizz] New Server Rendering Infra (#14144)

* [Fizz] Add Flow/Jest/Rollup build infra

Add a new package for react-stream which allows for custom server renderer
outputs. I picked the name because it's a reasonable name but also
because the npm name is currently owned by a friend of the project.

The react-dom build has its own inlined server renderer under the
name `react-dom/fizz`.

There is also a noop renderer to be used for testing. At some point
we might add a public one to test-renderer but for now I don't want to have
to think about public API design for the tests.

* Add FormatConfig too

We need to separate the format (DOM, React Native, etc) from the host
running the server (Node, Browser, etc).

* Basic wiring between Node, Noop and DOM configs

The Node DOM API is pipeToNodeStream which accepts a writable stream.

* Merge host and format config in dynamic react-stream entry point

Simpler API this way but also avoids having to fork the wrapper config.

Fixes noop builds.

* Add setImmediate/Buffer globals to lint config

Used by the server renderer

* Properly include fizz.node.js

Also use forwarding to it from fizz.js in builds so that tests covers
this.

* Make react-stream private since we're not ready to publish

or even name it yet

* Rename Renderer -> Streamer

* Prefix react-dom/fizz with react-dom/unstable-fizz

* Add Fizz Browser host config

This lets Fizz render to WHATWG streams. E.g. for rendering in a
Service Worker.

I added react-dom/unstable-fizz.browser as the entry point for this.

Since we now have two configurations of DOM. I had to add another
inlinedHostConfigs configuration called `dom-browser`. The reconciler
treats this configuration the same as `dom`. For stream it checks
against the ReactFizzHostConfigBrowser instead of the Node one.

* Add Fizz Browser Fixture

This is for testing server rendering - on the client.

* Lower version number to detach it from react-reconciler version

bvaughn committed:
Fixed scheduler setTimeout fallback (#14358)

* Fixed scheduler setTimeout fallback
* Moved unit-test-specific setTimeout code into a new NPM package, jest-mock-scheduler.

bvaughn committed:
Removed unnecessary externals from Jest bundles (#14372)

bvaughn committed:
Automated fixture tests (#14370)

* Renamed snapshot test from test.js to snapshot-test.js
* Automate fixtures tests

bvaughn committed:
Don't prompt to tag or create GitHub release for canary releases (#14376)

bvaughn committed:
Prevent a v8 deopt when profiling (#14383)

bvaughn committed:
Fix scheduler setTimeout() re-entrancy check (#14384)

AGCB committed:
fix spelling error: differen -> different (#14378)

Commit:
Tweaked wording for v8 "performance cliff" issue

sahrens committed:
Enable hooks in fabric (#14301)

because hooks are awesome.

bvaughn committed:
TestRenderer toJSON should not expose the Array wrapper Suspense uses for hidden trees (#14392)

bvaughn committed:
Added ErrorBoundary tests for useEffect and useLayoutEffect (#14401)

mrkev committed:
Implement pauseExecution, continueExecution, dumpQueue for Scheduler (#14053)

* Implement pauseExecution, continueExecution, dumpQueue

* Expose firstCallbackNode. Fix tests. Revert results.json

* Put scheduler pausing behind a feature flag

bvaughn committed:
Dangerfile exits early if build failed (#14400)

* Dangerfile exits early (without leaving an error comment) if build failed

bvaughn committed:
Updated version incrementing suggestion in release script based on team discussion (#14389)

NE-SmallTown committed:
Remove redundant argument of getPlugins function (#14419)

pleunv committed:
Add support for Suspense & lazy() to the react-is package (#14423)

* Add support for lazy & Suspense to react-is

bvaughn committed:
Inject overrideProps() fn to DevTools (#14427)

* Inject overrideProps() fn to DevTools

This function will enable editing props for function components, host nodes, and special types like memo and forwardRef.

bvaughn committed:
Enable hooks by default for FB React Native renderer (#14435)

* Enable hooks by default for FB React Native renderer
* Updated RN+FB feature flags to make some of the dynamic ones static

bvaughn committed:
Removed Fabric-specific feature flag files and updated Rollup to use the (non-Fabric) React Native flag files. (#14437)

acdlite committed:
Memoize promise listeners to prevent exponential growth (#14429)

* Memoize promise listeners to prevent exponential growth

Previously, React would attach a new listener every time a promise is
thrown, regardless of whether the same listener was already attached
during a previous render. Because React attempts to render every time
a promise resolves, the number of listeners grows quickly.

This was especially bad in synchronous mode because the renders that
happen when the promise pings are not batched together. So if a single
promise has multiple listeners for the same root, there will be multiple
renders, which in turn results in more listeners being added to the
remaining unresolved promises. This results in exponential growth in
the number of listeners with respect to the number of IO-bound
components in a single render.

Fixes #14220

* Memoize on the root and Suspense fiber instead of on the promise

* Add TODO to fix persistent mode tests

acdlite committed:
Make scheduler debugging feature flag static

acdlite committed:
Create separate SchedulerFeatureFlags instead of using ReactFeatureFlags (#14455)

acdlite committed:
Move SchedulerFeatureFlags fork to src directory to fix lint

acdlite committed:
React v16.7.0

acdlite committed:
Update CHANGELOG for 16.7

CarlMungazi committed:
Update ReactFiberScheduler.js (#14477)

Fixed typo

iljadaderko committed:
Allow node 11 as devEngine (#14450)

Using node 11 throws incompatibility error. This fixes it. Only test I performed was to run `yarn build -- --type=RN_OSS` everything seemed ok.

CarlMungazi committed:
Update ReactUpdateQueue.js (#14521)

Fix comment typo

sophiebits committed:
Simplify wording of key warning (#14503)

I don't think "array or iterator" is adding anything, and it may well be confusing, especially since this is one of the first and most common warnings that devs see.

mmarkelov committed:
Separate Object.is polyfill (#14334)

* Separate_Object_Is_Polyfill

bvaughn committed:
Separate current owner and dispatcher (#14548)

bvaughn committed:
Inject ReactCurrentDispatcher ref to DevTools (#14550)

Commit:
rename useImperativeMethods -> useImperativeHandle (#14565)

gaearon committed:
Support useLayoutEffect

gaearon committed:
Add some failing tests and comments

bvaughn committed:
react-debug-tools accepts currentDispatcher ref as param (#14556)

* react-debug-tools accepts currentDispatcher ref as param

* ReactDebugHooks injected dispatcher ref is optional

gaearon committed:
Gather dependencies in child scopes too

gaearon committed:
If we don't find foo.bar.baz in deps, try foo.bar, then foo

gaearon committed:
foo is enough for both foo.bar and foo.baz

CarlMungazi committed:
Fix typo (#14576)

Commit:
make a fork for ReactCurrentDispatcher (#14588)

bvaughn committed:
Support configurable labels for custom hooks (#14559)

* react-debug-tools accepts currentDispatcher ref as param

* ReactDebugHooks injected dispatcher ref is optional

* Support custom values for custom hooks

* PR feedback:

1. Renamed useDebugValueLabel hook to useDebugValue
2. Wrapped useDebugValue internals in if-DEV so that it could be removed from production builds.

* PR feedback:

1. Fixed some minor typos
2. Added inline comment explaining the purpose of  rollupDebugValues()
3. Refactored rollupDebugValues() to use a for loop rather than filter()
4. Improve check for useDebugValue hook to lessen the chance of a false positive
5. Added optional formatter function param to useDebugValue

* Nitpick renamed a method

aweary committed:
Avoid new Set([iterable]) for thenables (#14592)

Fixes https://github.com/facebook/react/issues/14583

Using `new Set([iterable])` does not work with IE11's non-compliant Set
implementation. By avoiding this pattern we don't need to require a Set
polyfill for IE11

aweary committed:
Tweak to avoid property read (#14593)

bvaughn committed:
Rename release script fixture test scheduler->tracing (#14590)

bvaughn committed:
Add noop useDebugValue hook to partial/server renderer (#14597)

bvaughn committed:
useDebugValue should throw if used in a class component (#14601)

bvaughn committed:
Added Flow type to keep hooks dispatchers in-sync (#14599)

* Added Flow type to keep hooks dispatchers in-sync

gaearon committed:
Add ESLint rule playground (#14609)

* Add ESLint rule playground

* Update index.js

* Update index.js

gaearon committed:
Shorter rule name

AdeelMalik22 reviewed (commented):
hi there
